rcljava/src/main/java/org/ros2/rcljava/action/ActionServerImpl.java

```diff
@@ -0,0 +1,561 @@
+    this.handle = nativeCreateActionServer(
+      node.getHandle(), node.getClock().getHandle(), actionType, actionName);
+    // TODO(jacobperron): Introduce 'Waitable' interface for entities like timers, services, etc
+    // node.addWaitable(this);
+  }
+
+  private static native int nativeGetNumberOfSubscriptions(long handle);
+  private static native int nativeGetNumberOfTimers(long handle);
+  private static native int nativeGetNumberOfClients(long handle);
+  private static native int nativeGetNumberOfServices(long handle);
+
+  /**
+   * {@inheritDoc}
+   */
+  public int getNumberOfSubscriptions() {
+    return nativeGetNumberOfSubscriptions(this.handle);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public int getNumberOfTimers() {
+    return nativeGetNumberOfTimers(this.handle);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public int getNumberOfClients() {
+    return nativeGetNumberOfClients(this.handle);
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public int getNumberOfServices() {
+    return nativeGetNumberOfServices(this.handle);
+  }
+
+  private static native boolean[] nativeGetReadyEntities(
+    long actionServerHandle, long waitSetHandle);
+
+  /**
+   * {@inheritDoc}
+   */
+  public boolean isReady(long waitSetHandle) {
+    this.readyEntities = nativeGetReadyEntities(this.handle, waitSetHandle);
+    for (boolean isReady : this.readyEntities) {
+      if (isReady) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private ActionServerGoalHandle<T> executeGoalRequest(
+    RMWRequestId rmwRequestId,
+    GoalRequestDefinition requestMessage,
+    GoalResponseDefinition responseMessage)
+  {
+    builtin_interfaces.msg.Time timeRequestHandled = this.clock.now().toMsg();
+    responseMessage.setStamp(timeRequestHandled.getSec(), timeRequestHandled.getNanosec());
+
+    // Create and populate a GoalInfo message
+    List<Byte> goalUuid = requestMessage.getGoalUuid();
+    action_msgs.msg.GoalInfo goalInfo = new action_msgs.msg.GoalInfo();
+    unique_identifier_msgs.msg.UUID uuidMessage= new unique_identifier_msgs.msg.UUID();
+    uuidMessage.setUuid(goalUuid);
+    goalInfo.setGoalId(uuidMessage);
+    goalInfo.setStamp(timeRequestHandled);
+
+    long goalInfoFromJavaConverterHandle = goalInfo.getFromJavaConverterInstance();
+    long goalInfoDestructorHandle = goalInfo.getDestructorInstance();
+
+    // Check that the goal ID isn't already being used
+    boolean goalExists = nativeCheckGoalExists(
+      this.handle, goalInfo, goalInfoFromJavaConverterHandle, goalInfoDestructorHandle);
+    if (goalExists) {
+      logger.warn("Received goal request for goal already being tracked by action server. Goal ID: " + goalUuid);
+      responseMessage.accept(false);
+      return null;
+    }
+
+    // Workaround type
+    GoalCallback<GoalRequestDefinition> callback = ((ActionServerImpl) this).goalCallback;
+    // Call user callback
+    GoalCallback.GoalResponse response = callback.handleGoal(requestMessage);
```

Comment:
I'm surprised java allows you to do this, as it's not type safe.

I was entertaining the idea of doing something like:

```java
interface GoalRequestDefinition<Action extends ActionDefinition> extends MessageDefinition {...}
```
```java
class ActionServerImpl<Action extends ActionDefinition, GoalRequest extends GoalRequestDefinition, ...> implements ActionServer<Action extends ActionDefinition, GoalRequest extends GoalRequestDefinition, ...>
```

```java
actionServer = new ActionServer<MyAction, MyAction.GoalRequest, ...>(...);
```
in this way it's impossible to pass a callback with the wrong type.

See also https://github.com/osrf/ros2_java/pull/52/files#r551441871.

Reply:
I looked into what's going on here. [This SO post is informative](https://stackoverflow.com/questions/3546745/multiple-wildcards-on-a-generic-methods-makes-java-compiler-and-me-very-confu). 

Basically, we have a type mismatch here already. `requestMessage` has type `GoalRequestDefinition<T>` and `goalCallback.handleGoal()` expects type `? extends GoalRequestDefinition<T>`. These are simply not compatible.

With most recent two changes 77d8699 and 21b194a, I think this code safer. Even though we're using type-erasure here as a workaround, the user must still provide a callback operating on a `GoalRequestDefinition<T extends ActionDefinition>`. Perhaps adding more template parameters as you suggest will also work, though it does seem awfully verbose..

Reply:
> With most recent two changes 77d8699 and 21b194a, I think this code is safer. Even though we're using type-erasure here as a workaround, the user must still provide a callback operating on a GoalRequestDefinition<T extends ActionDefinition>. Perhaps adding more template parameters as you suggest will also work, though it does seem awfully verbose..

I agree that more templates parameters would be awfully verbose, and with the latest changes things look better as there is only one generated class that implements GoalRequestDefinition<T>.
So I think this now looks fine